### PR TITLE
[FIX] [REFACTOR] FTP crawler

### PR DESCRIPTION
**Pull request by mathdugre**
## Description

When there was an "[Err 104] Connection reset by peer" the script was hanging.
Also improved the overall performance and use git api instead of OS call.

### CODE REVIEW

File: scripts/datalad_crawlers/ftp_crawler.py [hunk 1]
```diff
@@ -2,24 +2,37 @@
...
+                elif metadata.get("type") == "file":
+                    git.Repo().git.annex(
+                        "addurl",
+                        f"ftp://{host}/" + os.path.join(cwd, filename),
+                        file=filepath,
+                    )
+                break
+
+            except Exception as e:
+                print(f"WARNING: Connection restarted on file: {filepath}: {e}")
```
Comment (cmadjar) [inline scripts/datalad_crawlers/ftp_crawler.py:33]: I see that the FTP connection is reset in here but if the connection failed on the file, doesn't it mean that this file is skipped and it goes to the next filename?

**Comment by cmadjar** [comment]
@mathdugre One question/comment above. Thank you for submitting the changes to the script!!

**Comment by mathdugre** [comment]
@cmadjar , we might want to wait before merging the PR. I'm still experiencing some issue when crawling a large dataset; if there's an internet problem on the user side, the user will have to restart  the crawling from scratch.
I'm currently implementing the feature to crawl only a specific file/sub-directory of the dataset.

**Comment by mathdugre** [comment]
@cmadjar, I implemented the feature to crawl subdirectories. The PR will be ready for review again.

File: scripts/datalad_crawlers/ftp_crawler.py (hunk 1 above)
Comment (cmadjar) [reply scripts/datalad_crawlers/ftp_crawler.py:33]: @mathdugre did you answer my question above? Sorry if I don't remember. I don't see anything in the comments of the PR as far as I can tell.

**Comment by cmadjar** [comment]
@mathdugre see question above I had: https://github.com/CONP-PCNO/conp-dataset/pull/637#discussion_r646890211

**Comment by cmadjar** [comment]
@mathdugre you will also need to merge master into your branch to fix some of the Circle CI errors.

**Comment by mathdugre** [comment]
> @mathdugre see question above I had: [#637 (comment)](https://github.com/CONP-PCNO/conp-dataset/pull/637#discussion_r646890211)

Not sure what happened with my answer, sorry for that.

The crawler will only proceed to the next file it succeed to download the current one. That is, it will retry the same file again and again until it is done downloading the file/folder; see line 19 and 30.
Do you think it would be good to have a maximum amount of tries? If so, should the crawler stop after a maximum retry failure? Or proceed with the next file and log the error in a file for the user to see?

**Comment by cmadjar** [comment]
@mathdugre I guess the WARNING will keep showing as long as the file cannot be downloaded. If so, I guess the user can decide to cancel whenever they want. Maybe it is best to let it run, just in case the connection loss is momentary?

Would that make sense leaving it running and letting the user stop the run if they think it is taking too long? If not, I would just end the crawling with the error message displayed in the terminal.

For sure, I would not proceed to the next file as there are big chances that the failure will be missed in case someone forgets to look at the log.

**Comment by mathdugre** [comment]
@cmadjar, I agree that we should not skip a file that was not downloaded, as the user can miss the warning.
I think having the script run until it can download the file should be fine. If some issues arise with time, we can change it to a max attempts before exiting with error.

**Review by cmadjar — APPROVED** [review]
LGTM. Thank you @mathdugre !

**Pull request merged**